### PR TITLE
announce payment threshold through pricing protocol

### LINKED ISSUE #417: Payment threshold and disconnects
# Task

Instruct settlement to do a payment if threshold is met. Disconnect if disconnect threshold is exceeded. A mock settlement layer can be used here.

- [x] `Settlement` interface with a mock implementation where settlement always works
  - [x] Create a simple protocol with a pay message containing the value
  - [x] Adjust accounting balance on received pay message
- [x] Introduce accounting/pricing protocol
  - [x] Communicate payment threshold to connected peers
  - [x] Introduce per-peer payment threshold (their threshold)
  - [x] Adjust peers payment threshold on received message
- [x] Changes to `Accounting`
  - [x] Introduce a fixed payment threshold (for now)
  - [x] Call `Settlement` if payment threshold exceeded or if we are close (depends on details of `Reserve`)
  - [x] Adjust `Reserve` to work with peers payment threshold
  - [x] Return error in `Add` if disconnect threshold is met
  - [x] Instead of configurable disconnect threshold, use payment treshold plus tolerance

## Acceptance criterea

Peers exchange payment threshold and pay each other with fake payment to restore balance.

### PR DESCRIPTION
**Pull request by ralph-pichler**
closes #417

### CODE REVIEW

**Comment by Eknir** [comment]
@ralph-pichler , given that this one is still in draft, I assume you are not aiming to have it in for the next release, right?

**Comment by ralph-pichler** [comment]
> @ralph-pichler , given that this one is still in draft, I assume you are not aiming to have it in for the next release, right?

yes, since this depends on some yet to do changes in the protocol layer (init function) I would not include it in the release.

**Comment by janos** [comment]
> > @ralph-pichler , given that this one is still in draft, I assume you are not aiming to have it in for the next release, right?
> 
> yes, since this depends on some yet to do changes in the protocol layer (init function) I would not include it in the release.

Yes, sorry for that. There was some short discussion on the core channel https://beehive.ethswarm.org/swarm/pl/isjam1ro9fnpjfssi7d61p6scc about the Init function in ProtocolSpec, but not a final conclusion if it is a good approach. I know that @pradovic liked the idea, but also would like to her if @acud would approve that. Somehow this was left unresolved. @acud would it be ok to proceed with the architectural proposal on Init function in ProtocolSpec? I am asking you mainly as you implemented topology.Notifier interface which is similar but does notifications only when a peer connects to node, not when the node makes a connection to a peer, as Connect is used and not ConnectNotify.

File: pkg/accounting/accounting.go [hunk 1]
```diff
@@ -60,6 +60,7 @@ type Options struct {
 	Logger           logging.Logger
 	Store            storage.StateStorer
 	Settlement       settlement.Interface
+	Pricing          pricing.Interface
```
Comment (pradovic) [inline pkg/accounting/accounting.go:63]: I would move all fields out of the Options struct that are not really optional.

File: pkg/accounting/accounting.go [hunk 2]
```diff
@@ -495,3 +496,26 @@ func addI64pU64(a int64, b uint64) (result int64, err error) {
...
+
+	accountingPeer.lock.Lock()
+	defer accountingPeer.lock.Unlock()
+
+	accountingPeer.paymentThreshold = paymentThreshold
+	return nil
+}
+
+// Connected is called when a peer connects to us
+func (a *Accounting) Connected(ctx context.Context, peer swarm.Address) error {
```
Comment (pradovic) [inline pkg/accounting/accounting.go:515]: How is this going to be triggered?

File: pkg/accounting/accounting.go [hunk 3]
```diff
@@ -55,13 +55,6 @@ type accountingPeer struct {
 
 // Options are options provided to Accounting.
 type Options struct {
```
Comment (ralph-pichler) [inline pkg/accounting/accounting.go:57]: if it's empty drop this entirely

File: pkg/accounting/accounting_test.go [hunk 4]
```diff
@@ -662,3 +599,78 @@ func TestAccountingInvalidPaymentTolerance(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = acc.NotifyPaymentThreshold(peer1Addr, testPaymentThreshold)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO: check the success somehow
```
Comment (pradovic) [inline pkg/accounting/accounting_test.go:675]: todo?

File: pkg/node/node.go [hunk 5]
```diff
@@ -282,19 +283,15 @@ func NewBee(addr string, swarmAddress swarm.Address, keystore keystore.Service,
...
 
-	acc, err := accounting.NewAccounting(accounting.Options{
-		Logger:           logger,
-		Store:            stateStore,
-		PaymentThreshold: o.PaymentThreshold,
-		PaymentTolerance: o.PaymentTolerance,
-		EarlyPayment:     o.PaymentEarly,
-		Settlement:       settlement,
-	})
+	pricing := pricing.New(p2ps, logger, o.PaymentThreshold)
```
Comment (pradovic) [inline pkg/node/node.go:286]: Should `p2ps.AddProtocol()` be called for pricing as well?

**Review by pradovic — COMMENTED** [review]
Nice work, code looks very nice. Left some questions about some stuff that might be missing, if I am not wrong.

**Review by paxthemax — APPROVED** [review]
Code looks good to go as far as I'm concerned.

File: pkg/node/node.go (hunk 5 above)
Comment (ralph-pichler) [reply pkg/node/node.go:294]: added

File: pkg/accounting/accounting_test.go (hunk 4 above)
Comment (ralph-pichler) [reply pkg/accounting/accounting_test.go]: test completed

File: pkg/accounting/accounting.go (hunk 1 above)
Comment (ralph-pichler) [reply pkg/accounting/accounting.go]: removed

File: pkg/accounting/accounting.go (hunk 2 above)
Comment (ralph-pichler) [reply pkg/accounting/accounting.go]: removed

**Review by pradovic — APPROVED** [review]
👍

**Pull request merged**